src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java

```diff
@@ -83,39 +85,46 @@ public FavoriteDTO findFavoriteById(UUID id) {
 
     @Transactional
     @Override
-    public FavoriteDTO assignFavoriteToUser(CreateFavoriteDTO createFavoriteDTO) {
-        log.debug("Finding of certain user by ID");
+    public FavoriteDTO assignFavoriteToUser(UUID discountId) {
+        log.debug(String.format("Check: if user already have Favorite of Discount with ID %s ", discountId));
+        String userEmail = SecurityContextHolder.getContext().getAuthentication().getName();
+        if (favoriteRepository.findByDiscountIdAndAndUserEmail(discountId, userEmail).isPresent()) {
+            throw  new CreationRestrictedException(String
+                    .format("Favorite for discount with id %s does already exist", discountId));
+        }
+
         User user = userRepository
-                .findById(createFavoriteDTO.getUserId())
+                .findByEmail(userEmail)
                 .orElseThrow(() -> new NotFoundException(String
-                        .format("User with id %s not found", createFavoriteDTO.getUserId())));
+                        .format("User with email %s not found", userEmail)));
 
         Discount discount = discountRepository
-                .findById(createFavoriteDTO.getDiscountId())
+                .findById(discountId)
                 .orElseThrow(() -> new NotFoundException(String
-                        .format("Discount with id %s not found", createFavoriteDTO.getDiscountId())));
-        log.debug("Successfully certain User and Discount are found by ID. Starting Favorite creation/saving.");
+                        .format("Discount with id %s not found", discountId)));
+        log.debug("Successfully certain User and Discount are found by ID. Starting favorite creation/saving.");
 
         Favorite favorite = new Favorite();
         favorite.setUser(user);
         favorite.setDiscount(discount);
-
         Favorite favoriteSaved = favoriteRepository.save(favorite);
-        log.debug("Successfully new Favorite is saved to certain user");
+        log.debug(String.format("Successfully added Favorite for Discount with ID %s to user", discountId));
 
         return favoriteMapper.toFavoriteDTO(favoriteSaved);
     }
 
     @Transactional
     @Override
-    public void deleteFavoriteByID(UUID id) {
-        log.debug("Finding & deleting Favorite by ID");
+    public void deleteFavoriteByID(UUID discountId) {
+        log.debug("Finding & deleting Favorite by DiscountID");
+        String userEmail = SecurityContextHolder.getContext().getAuthentication().getName();
+        if (favoriteRepository.findByDiscountIdAndAndUserEmail(discountId, userEmail).isEmpty()) {
+            throw new DeletionRestrictedException(String
+                    .format("Favorite for discount with id %s does not already exist", discountId));
+        }
```

Comment:
You can use existsByDiscountIdAndUserEmail, it'll return boolean

Reply:
I do not know why - it throw me wierd exaption abot not presend fetch operation owner.
Let leave it for a whole for a speed, please

Reply:
I've checked - it seems that everything works. Just use correct naming and remove Entity Graph